### PR TITLE
[SG2] adding Slider node and Custom Texture Render nodes to the staging branch

### PR DESCRIPTION
**Pull request by tracychen1234**
---
### Purpose of this PR
Add below nodes to node library: 
1. Slider
2. Custom Render Texture Self
3. Custom Render Texture Size
4. Slice Index/Cubemap Face

---
### Testing status
Created nodes in an sg2 graph with no error thrown

---

### CODE REVIEW

File: com.unity.sg2/Editor/GraphDeltaRegistry/FunctionDefinitions/StandardDefinitions/Utility/CustomRenderTexture/CustomRenderTextureSizeNode.cs [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+{//TODO: define checkings for preview shader
+    internal class CustomRenderTextureSizeNode : IStandardNode
+    {
+        static string Name = "CustomRenderTextureSize";
+        static int Version = 1;
+        public static FunctionDescriptor FunctionDescriptor => new(
+            Version,
+            Name,
+@"
+TextureHeight = CustomRenderTextureHeightRef;
```
Comment (bencloward) [inline com.unity.sg2/Editor/GraphDeltaRegistry/FunctionDefinitions/StandardDefinitions/Utility/CustomRenderTexture/CustomRenderTextureSizeNode.cs:15]: Nit - can we put these in the same order as they appear on the node? (Width, Height, Depth)

File: com.unity.sg2/Editor/GraphDeltaRegistry/FunctionDefinitions/StandardDefinitions/Utility/CustomRenderTexture/CustomRenderTextureSizeNode.cs [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+TextureDepth = CustomRenderTextureDepthRef;
```
Comment (bencloward) [inline com.unity.sg2/Editor/GraphDeltaRegistry/FunctionDefinitions/StandardDefinitions/Utility/CustomRenderTexture/CustomRenderTextureSizeNode.cs:16]: Here, it's called "CustomRenderTextureDepthRef" but in the params it's called TextureDepthRef.  They need to match.

File: com.unity.sg2/Editor/GraphDeltaRegistry/FunctionDefinitions/StandardDefinitions/Utility/CustomRenderTexture/CustomRenderTextureSizeNode.cs [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+            new ParameterDescriptor("TextureDepthRef", TYPE.Float, GraphType.Usage.Static),//TODO: need to be a ref texture type
+            new ParameterDescriptor("TextureWidth", TYPE.Float, GraphType.Usage.Out),
+            new ParameterDescriptor("TextureHeight", TYPE.Float, GraphType.Usage.Out),
+            new ParameterDescriptor("TextureDepth", TYPE.Float, GraphType.Usage.Out)
+        );
+
+        public static NodeUIDescriptor NodeUIDescriptor => new(
+            Version,
+            Name,
+            tooltip: "Return the height, depth and width of the custom texture.",
```
Comment (bencloward) [inline com.unity.sg2/Editor/GraphDeltaRegistry/FunctionDefinitions/StandardDefinitions/Utility/CustomRenderTexture/CustomRenderTextureSizeNode.cs:30]: Let's make this match our order as well - Width, Height, Depth

**Review by bencloward — CHANGES_REQUESTED** [review]

File: com.unity.sg2/Editor/GraphDeltaRegistry/FunctionDefinitions/StandardDefinitions/Utility/CustomRenderTexture/CustomRenderTextureSelfNode.cs [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+SelfTexture2D = SelfTexture2DRef;
+SelfTextureCube = SelfTextureCubeRef;
+SelfTexture3D = SelfTexture3DRef;
+",
+            new ParameterDescriptor("SelfTexture2DRef", TYPE.Vec4, GraphType.Usage.Static),//TODO: need to be a ref texture type
+            new ParameterDescriptor("SelfTextureCubeRef", TYPE.Vec4, GraphType.Usage.Static),//TODO: need to be a ref texture type
+            new ParameterDescriptor("SelfTexture3DRef", TYPE.Vec4, GraphType.Usage.Static),//TODO: need to be a ref texture type
+            new ParameterDescriptor("SelfTexture2D", TYPE.Vec4, GraphType.Usage.Out),
+            new ParameterDescriptor("SelfTextureCube", TYPE.Vec4, GraphType.Usage.Out),
+            new ParameterDescriptor("SelfTexture3D", TYPE.Vec4, GraphType.Usage.Out)
```
[the remark this answers is not in the document]
Comment (tracychen1234) [reply com.unity.sg2/Editor/GraphDeltaRegistry/FunctionDefinitions/StandardDefinitions/Utility/CustomRenderTexture/CustomRenderTextureSelfNode.cs:24]: Good call!

File: com.unity.sg2/Editor/GraphDeltaRegistry/FunctionDefinitions/StandardDefinitions/Utility/CustomRenderTexture/CustomRenderTextureSelfNode.cs [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+        public static NodeUIDescriptor NodeUIDescriptor => new(
+            Version,
+            Name,
+            tooltip: "Return a custom render texture.",
+            categories: new string[] { "Utility", "Custom Render Texture" },
+            synonyms: new string[0] { },
+            hasPreview: false,
+            parameters: new ParameterUIDescriptor[] {
+                new ParameterUIDescriptor(
+                    name: "SelfTexture2D",
```
[the remark this answers is not in the document]
Comment (tracychen1234) [reply com.unity.sg2/Editor/GraphDeltaRegistry/FunctionDefinitions/StandardDefinitions/Utility/CustomRenderTexture/CustomRenderTextureSelfNode.cs:36]: I thought Brett would be doing some auto-formatting to insert space in each word base on the name so we don't need to specify some names?

File: com.unity.sg2/Editor/GraphDeltaRegistry/FunctionDefinitions/StandardDefinitions/Utility/CustomRenderTexture/CustomRenderTextureSizeNode.cs [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+        static int Version = 1;
+        public static FunctionDescriptor FunctionDescriptor => new(
+            Version,
+            Name,
+@"
+TextureHeight = CustomRenderTextureHeightRef;
+TextureDepth = CustomRenderTextureDepthRef;
+TextureWidth = CustomRenderTextureWidthRef;
+",
+            new ParameterDescriptor("CustomRenderTextureHeightRef", TYPE.Float, GraphType.Usage.Static),//TODO: need to be a ref texture type
```
[the remark this answers is not in the document]
Comment (tracychen1234) [reply com.unity.sg2/Editor/GraphDeltaRegistry/FunctionDefinitions/StandardDefinitions/Utility/CustomRenderTexture/CustomRenderTextureSizeNode.cs:19]: ah good catch!

File: com.unity.sg2/Editor/GraphDeltaRegistry/FunctionDefinitions/StandardDefinitions/Utility/CustomRenderTexture/CustomRenderTextureSizeNode.cs [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+            new ParameterDescriptor("CustomRenderTextureWidthRef", TYPE.Float, GraphType.Usage.Static),//TODO: need to be a ref texture type
+            new ParameterDescriptor("TextureDepthRef", TYPE.Float, GraphType.Usage.Static),//TODO: need to be a ref texture type
```
[the remark this answers is not in the document]
Comment (tracychen1234) [reply com.unity.sg2/Editor/GraphDeltaRegistry/FunctionDefinitions/StandardDefinitions/Utility/CustomRenderTexture/CustomRenderTextureSizeNode.cs:21]: fixed!

File: com.unity.sg2/Editor/GraphDeltaRegistry/FunctionDefinitions/StandardDefinitions/Utility/CustomRenderTexture/CustomRenderTextureSizeNode.cs (hunk 2 above)
Comment (tracychen1234) [reply com.unity.sg2/Editor/GraphDeltaRegistry/FunctionDefinitions/StandardDefinitions/Utility/CustomRenderTexture/CustomRenderTextureSizeNode.cs:16]: fixed

File: com.unity.sg2/Editor/GraphDeltaRegistry/FunctionDefinitions/StandardDefinitions/Utility/CustomRenderTexture/CustomRenderTextureSelfNode.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (tracychen1234) [reply com.unity.sg2/Editor/GraphDeltaRegistry/FunctionDefinitions/StandardDefinitions/Utility/CustomRenderTexture/CustomRenderTextureSelfNode.cs:36]: fixed

File: com.unity.sg2/Editor/GraphDeltaRegistry/FunctionDefinitions/StandardDefinitions/Utility/CustomRenderTexture/CustomRenderTextureSizeNode.cs [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+        public static NodeUIDescriptor NodeUIDescriptor => new(
+            Version,
+            Name,
+            tooltip: "Return the height, depth and width of the custom texture.",
+            categories: new string[] { "Utility", "Custom Render Texture" },
+            synonyms: new string[0] { },
+            hasPreview: false,
+            parameters: new ParameterUIDescriptor[] {
+                new ParameterUIDescriptor(
+                    name: "TextureHeight",
```
[the remark this answers is not in the document]
Comment (tracychen1234) [reply com.unity.sg2/Editor/GraphDeltaRegistry/FunctionDefinitions/StandardDefinitions/Utility/CustomRenderTexture/CustomRenderTextureSizeNode.cs:36]: fixed

File: com.unity.sg2/Editor/GraphDeltaRegistry/FunctionDefinitions/StandardDefinitions/Utility/CustomRenderTexture/CustomRenderTextureSliceNode.cs [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+            Version,
+            Name,
+            tooltip: "Return the indexes of the current cubemap face or the current 3d slice.",
+            categories: new string[] { "Utility", "Custom Render Texture" },
+            synonyms: new string[0] {},
+            displayName: "Slice Index / Cubemap Face",
+            hasPreview: false,
+            parameters: new ParameterUIDescriptor[] {
+                new ParameterUIDescriptor(
+                    name: "TextureCubeFace",
```
[the remark this answers is not in the document]
Comment (tracychen1234) [reply com.unity.sg2/Editor/GraphDeltaRegistry/FunctionDefinitions/StandardDefinitions/Utility/CustomRenderTexture/CustomRenderTextureSliceNode.cs:34]: added

File: com.unity.sg2/Editor/GraphDeltaRegistry/FunctionDefinitions/StandardDefinitions/Utility/CustomRenderTexture/CustomRenderTextureSliceNode.cs [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+        static string Name = "CustomRenderTextureSlice";
+        static int Version = 1;
+        public static FunctionDescriptor FunctionDescriptor => new(
+            Version,
+            Name,
+@"
+TextureCubeFace = CustomRenderTextureCubeFaceRef;
+Texture3DSlice = CustomRenderTexture3DSliceRef;
+",
+            new ParameterDescriptor("CustomRenderTextureCubeFaceRef", TYPE.Float, GraphType.Usage.Static),//TODO: need to be a ref texture type
```
[the remark this answers is not in the document]
Comment (tracychen1234) [reply com.unity.sg2/Editor/GraphDeltaRegistry/FunctionDefinitions/StandardDefinitions/Utility/CustomRenderTexture/CustomRenderTextureSliceNode.cs:18]: yes, these nodes are only for the Custom Render Texture target. I added a TODO about they are target-specific nodes and need a validation msg. I imagine the system will handle target validation and pop the validation msg for nodes?

**Review by bencloward — APPROVED** [review]

**Review by brett-groshong — APPROVED** [review]

**Pull request merged**